auto-config received after the backhaul is operational

rmelotte opened the pull request:
This is a quick attempt at fixing https://github.com/prplfoundation/prplMesh/issues/886

Although it fixes the particular issue we were facing in using the Broadcom controller as part of the test `MAP-4.8.1_ETH_FH24G`, I'd like to know more about the implication of the change in commit `3072fe02`, so please comment on it.

In particular:
- do you see any potential issue with this specific change
- why can't we call `finalize_slaves_connect_state(true, soc)` unconditionally when receiving an autoconfig-reply?

rmelotte committed:
agent: backhaul_manager: refactor tlvSupportedFreqBand handling

The handling of the 2.4GHz and 5GHz bands were exactly the same,
except for the log messages and freq_type.

Since there are already 3 different frequency bands to handle, use a
switch-case to handle each of them.

If the band is supported, assign a band name and a freq_type. If it's
not supported, log it and return.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
agent: backhaul_manager: finalize each slave on autoconfig response

Since commit 17ad041b5eb40b790d08d0e8ba225fe829d134e0 when an
autoconfig response was received for a specific band,
finalize_slaves_connect_state() would only be called if the backhaul
was already operational on slave registration. This is because if it
wasn't operational at registration the next state change to
OPERATIONAL would have already called finalize_slaves_connect_state().

There is a case where finalize_slaves_connect_state() wouldn't be
called for all the slaves though: if the auto-config message for one
of the slave arrived after the FSM went to OPERATIONAL state.

In that case, the FSM state change would cause the interface for which
the auto-config reply wasn't received to be skipped with the following
message: "controller not discovered, skipping...".
The finalize_slaves_connect_state() function would then never be
called, and the slave wouldn't be operational.

Do not call finalize_slaves_connect_state() when the state changes to
operational. Instead, call it unconditionally when receiving an
autoconfig response.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte commented:
There is a sample 4.8.1 test here (with the broadcom controller): https://gitlab.com/prpl-foundation/prplMesh/-/jobs/451251833

morantr commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -2276,22 +2276,25 @@ bool ap_wlan_hal_dwpal::process_dwpal_event(char *buffer, int bufLen, const std:
-        if (dwpal_string_to_struct_parse(buffer, bufLen, fieldsToParse, sizeof(MACAddress)) ==
-            DWPAL_FAILURE) {
+        if (dwpal_string_to_struct_parse(buffer, bufLen, fieldsToParse,
+                                         sizeof(fieldsToParse)) == DWPAL_FAILURE) {
```
The last argument defined on `dwpal_string_to_struct_parse` from dwpal lib is defined with the following description:
`size_t userBufLen - The user total buffer size to hold parsed info from inside fieldsToParse`

This means it should have the sum of the sizes of vap_name, MACAddress, status_code.
I saw this example on dwpal code:
![image](https://user-images.githubusercontent.com/10611984/75431136-11908e80-5955-11ea-8212-32d95b12199e.png)

I have no idea why it is implemented like this, but as I can see we use this argument wrong on all dwpal code.
@kantera800 @tomereli

arnout replied:
If you look to the piece of INTERCAL code that is dwpal, you'll see that the `field` member of the fieldsToParse argument is assumed to point to a struct member of the first element of an array-of-structs, and an intricate way to avoid using the offsetof operator. The userBufLen argument is supposed to be th total size of that array-of-structs. So indeed, if you're not using an array-of-structs, you have to manually add up the size of the different fields.

tomereli replied:
I saw the same today with Denis and we also didn't get why we pass a partial size.

morantr commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -2305,6 +2308,10 @@ bool ap_wlan_hal_dwpal::process_dwpal_event(char *buffer, int bufLen, const std:
+        std::string mac;
+        beerocks::net::network_utils::linux_iface_get_mac(vap_name, mac);
+        LOG(DEBUG) << "BTM response source BSSID: " << mac;
+        msg->params.source_bssid = beerocks::net::network_utils::mac_from_string(mac);
```
You added to the struct also target_bssid, but I didn't see you fill it, also on other bwl hals.

morantr commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2225,10 +2225,11 @@ bool backhaul_manager::handle_1905_autoconfiguration_response(ieee1905_1::CmduMe
                     // This means that it was restarted and sending backhaul connected event
                     // is required so that the slave will re-join the controller.
-                    if (soc->operational_on_registration) {
+                    if (soc->operational_on_registration || FSM_IS_IN_STATE(OPERATIONAL)) {
```
This change reverts commit https://github.com/prplfoundation/prplMesh/commit/17ad041b5eb40b790d08d0e8ba225fe829d134e0.

The reason we shouldn't do it is that once we got the first discovery response message we would move to an operational state and send the "backhaul connected" message to all son_slave which are discovered.
This means that after we will get the second discovery response we would get inside the "finalize" function and send this message again. This means that the previous son_slave which we got the discovery response for, will receive a "backhaul connected" message twice.

rmelotte replied:
Ok, I didn't notice I was actually reverting a previous change.
Thanks for the explanation.
And what if we only sent the "backhaul connected" message only when we have the autoconfig response and not when moving to the operational state? Does it make any sense? 
I'll push the change to this branch so that you can see what I mean.

tomereli replied:
>This means that the previous son_slave which we got the discovery response for, will receive a "backhaul connected" message twice.

So? The slave can simply ignore the second message. This is merely an optimization, and this PR proves it was wrong since it didn't cover all cases.

I think that instead of letting Raphael struggle with a new solution we should revert the optimization and be done with it.

@arnout @vitalybu what do you think?

morantr replied:
This 

> > This means that the previous son_slave which we got the discovery response for, will receive a "backhaul connected" message twice.
> 
> So? The slave can simply ignore the second message. This is merely an optimization, and this PR proves it was wrong since it didn't cover all cases.
> 
> I think that instead of letting Raphael struggle with a new solution we should revert the optimization and be done with it.
> 
> @arnout @vitalybu what do you think?

This is actually bad because when the son_slave will receive this message it will change its state to a different state and send messages it already sent.
Don't forget, we both did this because we noticed it causing a mess on the son_slave and the logs too.
BTW, this is one of the reasons we need unified Agent and not two independent "son_slaves".

tomereli replied:
So we should add a check for that in the son slave - if it is already operational it can ignore the second message.
I know that this is probably not the best solution so if you have another one please speak up, we need this resolved **today**

rmelotte replied:
@tomereli @morantr could you comment on commit ce621129ae9adf2ff0acc822bd544e0adf291206 ? 
Why not call the finalize function when receiving the autoconfig reply only?
It avoids the duplicate M1 and shouldn't put the slave in a wrong state as far as I understand.

morantr replied:
I don't remember why we did it like this, but your suggestion might be a good idea.
I think that the logic behinds it is not to hold back auto-config flow if the other radio hasn't been answered, but basically I don't see any reason why would it happen.

arnout commented:
It took me a long time to understand what is going on, but I definitely like the direction in which this is going. Actually, I'd make a few additional modifications.

- Further refactor the finalize_slaves_connect_state, split it into two functions for the connect and disconnect branches. They're completely disjoint. Actually, since each function is called only once, they could be inlined, but they're rather big so let's keep them as functions.
- After the last commit, pSocket will always be non-nullptr, so the looping over the different sockets becomes redundant. I would then move the loop over socs that is currently in handle_1905_autoconfiguration_response down into finalize_slaves_connect_state, and just pass the freq_type as argument instead of pSocket.

tomereli commented:
Definitely looks good to me, you can change it from RFC to a normal PR and IMO you can just drop the WIP commit so we can merge it tomorrow and pass the test officially before the arch call.

rmelotte commented:
> * Further refactor the finalize_slaves_connect_state, split it into two functions for the connect and disconnect branches. They're completely disjoint. Actually, since each function is called only once, they could be inlined, but they're rather big so let's keep them as functions.
> 
> * After the last commit, pSock

Since we need to fix this issue ASAP, I opened https://github.com/prplfoundation/prplMesh/issues/911 with your additional suggestions.

rmelotte commented:
Rebased on master, squashed the commits and changed the commit messages.
A new `MAP-4.8.1_ETH_FH24G` test (broadcom controller) has been run here: https://gitlab.com/prpl-foundation/prplMesh/-/jobs/459082965 (passed)
